In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
%matplotlib inline

In [6]:
df = pd.read_csv('bank-additional-full.csv', sep=';')

In [7]:
# There are 101,766 rows and 50 columns
print(df.shape)

(41188, 21)


In [9]:
# Examine data, there are some hidden columns
# show all columns
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

In [10]:
# examine the common stats of numeric columns
print(df.describe())

               age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198    

In [12]:
# loop through the data of each column, find the column has value of cell which is object =='?' then count

for col in df.columns:
    if df[col].dtype == object:
         print(col, df[col][df[col] == '?'].count())

# There are 101,766 rows and 50 columns
print(df.shape)

job 0
marital 0
education 0
default 0
housing 0
loan 0
contact 0
month 0
day_of_week 0
poutcome 0
y 0
(41188, 21)


In [8]:
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [14]:
# Because there are many of hidden columns & hidden rows,
#in order to have an overview of dataframe, try to see through all of them
#there are a lot missing value columns & cells
pd.set_option('display.max_columns', None)
#pd.set_option('display.max_rows', None)
df

age            job   marital            education  default  housing  \
0       56      housemaid   married             basic.4y       no       no   
1       57       services   married          high.school  unknown       no   
2       37       services   married          high.school       no      yes   
3       40         admin.   married             basic.6y       no       no   
4       56       services   married          high.school       no       no   
5       45       services   married             basic.9y  unknown       no   
6       59         admin.   married  professional.course       no       no   
7       41    blue-collar   married              unknown  unknown       no   
8       24     technician    single  professional.course       no      yes   
9       25       services    single          high.school       no      yes   
10      41    blue-collar   married              unknown  unknown       no   
11      25       services    single          high.school       no      yes   
12      29    blue-collar    single          high.school       no       no   
13      57      housemaid  divorced             basic.4y       no      yes   
14      35    blue-collar   married             basic.6y       no      yes   
15      54        retired   married             basic.9y  unknown      yes   
16      35    blue-collar   married             basic.6y       no      yes   
17      46    blue-collar   married             basic.6y  unknown      yes   
18      50    blue-collar   married             basic.9y       no      yes   
19      39     management    single             basic.9y  unknown       no   
20      30     unemployed   married          high.school       no       no   
21      55    blue-collar   married             basic.4y  unknown      yes   
22      55        retired    single          high.school       no      yes   
23      41     technician    single          high.school       no      yes   
24      37         admin.   married          high.school       no      yes   
25      35     technician   married    university.degree       no       no   
26      59     technician   married              unknown       no      yes   
27      39  self-employed   married             basic.9y  unknown       no   
28      54     technician    single    university.degree  unknown       no   
29      55        unknown   married    university.degree  unknown  unknown   
...    ...            ...       ...                  ...      ...      ...   
41158   35     technician  divorced             basic.4y       no       no   
41159   35     technician  divorced             basic.4y       no      yes   
41160   33         admin.   married    university.degree       no       no   
41161   33         admin.   married    university.degree       no      yes   
41162   60    blue-collar   married             basic.4y       no      yes   
41163   35     technician  divorced             basic.4y       no      yes   
41164   54         admin.   married  professional.course       no       no   
41165   38      housemaid  divorced    university.degree       no       no   
41166   32         admin.   married    university.degree       no       no   
41167   32         admin.   married    university.degree       no      yes   
41168   38   entrepreneur   married    university.degree       no       no   
41169   62       services   married          high.school       no      yes   
41170   40     management  divorced    university.degree       no      yes   
41171   33        student   married  professional.course       no      yes   
41172   31         admin.    single    university.degree       no      yes   
41173   62        retired   married    university.degree       no      yes   
41174   62        retired   married    university.degree       no      yes   
41175   34        student    single              unknown       no      yes   
41176   38      housemaid  divorced          high.school       no      yes   
41177   57        retired   m

In [43]:
df.groupby(['poutcome', 'y']).mean()

age    duration  campaign       pdays  previous  \
poutcome    y                                                            
failure     no   39.626817  211.764464  2.028517  980.291198  1.139841   
            yes  41.871074  467.224793  1.864463  879.682645  1.342149   
nonexistent no   39.935167  221.565850  2.710567  999.000000  0.000000   
            yes  40.173193  623.688634  2.184973  999.000000  0.000000   
success     no   40.453027  241.175365  1.991649    5.624217  1.421712   
            yes  42.864653  363.680089  1.710291    5.568233  1.648770   

                 emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
poutcome    y                                                             
failure     no      -1.611489       93.052742     -43.337209   1.740383   
            yes     -2.131240       93.158255     -39.002645   1.112688   
nonexistent no       0.491993       93.672219     -40.300904   4.083050   
            yes     -0.808946       93.372055     -40.506304   2.669357   
success     no      -2.042589       93.165113     -39.477662   1.199499   
            yes     -2.117338       93.425035     -37.805034   0.887831   

                 nr.employed  
poutcome    y                 
failure     no   5099.794132  
            yes  5049.159504  
nonexistent no   5186.565289  
            yes  5125.858007  
success     no   5053.795198  
            yes  5018.206600

In [1]:
df.groupby(['job', 'y']).mean()

NameError: name 'df' is not defined

In [ ]:
for i in df['y']:
    if i == no

Input variables:
# bank client data:
1 - age (numeric)

2 - job : type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')

3 - marital : marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)

4 - education (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')

5 - default: has credit in default? (categorical: 'no','yes','unknown')

6 - housing: has housing loan? (categorical: 'no','yes','unknown')

7 - loan: has personal loan? (categorical: 'no','yes','unknown')
# related with the last contact of the current campaign:

8 - contact: contact communication type (categorical: 'cellular','telephone') 

9 - month: last contact month of year (categorical: 'jan', 'feb', 'mar', ..., 'nov', 'dec')

10 - day_of_week: last contact day of the week (categorical: 'mon','tue','wed','thu','fri')

11 - duration: last contact duration, in seconds (numeric). Important note: this attribute highly affects the output target (e.g., if duration=0 then y='no'). Yet, the duration is not known before a call is performed. Also, after the end of the call y is obviously known. Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.

# other attributes:

12 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)

13 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric; 999 means client was not previously contacted)

14 - previous: number of contacts performed before this campaign and for this client (numeric)

15 - poutcome: outcome of the previous marketing campaign (categorical: 'failure','nonexistent','success')

# social and economic context attributes

16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)

17 - cons.price.idx: consumer price index - monthly indicator (numeric) 

18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric) 

19 - euribor3m: euribor 3 month rate - daily indicator (numeric)

20 - nr.employed: number of employees - quarterly indicator (numeric)

Output variable (desired target):
21 - y - has the client subscribed a term deposit? (binary: 'yes','no')